File: flask/app/services/test_service.py
import csv
import os
from datetime import datetime
import tempfile
from zipfile import ZipFile

from werkzeug.utils import secure_filename

from app import db

class TestService:
    def __init__(self, user_repo, group_repo, usergroupstate_repo):
        self.user_repo = user_repo
        self.group_repo = group_repo
        self.usergroupstate_repo = usergroupstate_repo
        self.temp_files_by_original_filename = {}

    def handle_file_upload(self, file):
        try:
            self.open_zip_file(file)
            self.import_from_file()
        finally:
            self.cleanup_temp_files()

    def open_zip_file(self, file):
        try:
            with ZipFile(file) as zip_ref:
                for file_info in zip_ref.infolist():
                    if file_info.filename.endswith('.csv'):
                        with zip_ref.open(file_info) as extracted_file:
                            temp_file = self.create_temp_file(file_info.filename)
                            with open(temp_file, 'wb') as temp_out:
                                temp_out.write(extracted_file.read())
        except Exception as e:
            raise RuntimeError(f"Error extracting ZIP file: {e}")

    def create_temp_file(self, file_name):
        temp_file = tempfile.NamedTemporaryFile(delete=False, prefix='tmp_', suffix=f"_{secure_filename(file_name)}")
        self.temp_files_by_original_filename[file_name] = temp_file.name
        return temp_file.name

    def get_temp_file(self, file_name):
        return self.temp_files_by_original_filename.get(file_name)

    def cleanup_temp_files(self):
        for file_path in self.temp_files_by_original_filename.values():
            try:
                os.remove(file_path)
            except Exception as e:
                print(f"Failed to delete temp file {file_path}: {e}")
        self.temp_files_by_original_filename.clear()

    def import_from_file(self):
        self.insert_group_state()
        self.insert_group()
        self.insert_user()

    def insert_group_state(self):
        file_path = self.get_temp_file("usergroupstate.csv")
        if not file_path:
            return

        with open(file_path, mode='r', encoding='utf-8') as csv_file:
            csv_reader = csv.reader(csv_file)
            next(csv_reader)  # Skip header
            try:
                for line in csv_reader:
                    state = self.usergroupstate_repo(
                        id=int(line[0]),
                        label=line[1]
                    )
                    db.session.merge(state)
                db.session.commit()
            except Exception as e:
                db.session.rollback()
                raise RuntimeError(f"Error inserting group states: {e}")

    def insert_group(self):
        file_path = self.get_temp_file("group.csv")
        if not file_path:
            return

        with open(file_path, mode='r', encoding='utf-8') as csv_file:
            csv_reader = csv.reader(csv_file)
            next(csv_reader)  # Skip header
            try:
                for line in csv_reader:
                    group = self.group_repo(
                        id=int(line[0]),
                        name=line[1],
                        contract_signed=datetime.strptime(line[2], '%Y-%m-%d %H:%M:%S.%f'),
                        contract_number=int(line[3]),
                        institute_name=line[4],
                        institute_address_name=line[5],
                        institute_address_number=line[6],
                        institute_postal_code=line[7],
                        province=line[8],
                        city=line[9],
                        country=line[10],
                        vat=line[11],
                        institute_phone=line[12],
                        institute_email=line[13],
                        state_id=int(line[14]),
                        priority=float(line[16])
                    )
                    db.session.merge(group)
                db.session.commit()
            except Exception as e:
                db.session.rollback()
                raise RuntimeError(f"Error inserting groups: {e}")

    def insert_user(self):
        file_path = self.get_temp_file("user.csv")
        if not file_path:
            return

        with open(file_path, mode='r', encoding='utf-8') as csv_file:
            csv_reader = csv.reader(csv_file)
            next(csv_reader)  # Skip header
            try:
                for line in csv_reader:
                    user = self.user_repo(
                        id=int(line[0]),
                        group_id=int(line[1]),
                        state_id=int(line[2]),
                        name=line[3],
                        surname=line[4],
                        login=line[5],
                        passwd=line[6],
                        email=line[7],
                        phone=line[8],
                        is_dsm_admin=line[9].lower() == 'true',
                        is_group_admin=line[10].lower() == 'true',
                        is_planner=line[11].lower() == 'true',
                        is_support_tech=line[12].lower() == 'true',
                        photo=line[13]
                    )
                    db.session.merge(user)
                db.session.commit()
            except Exception as e:
                db.session.rollback()
                raise RuntimeError(f"Error inserting users: {e}")

File: express/services/testService.js
const fs = require('fs');
const os = require('os');
const path = require('path');
const {promisify} = require('util');
const {v4: uuidv4} = require('uuid');
const {pipeline} = require('stream');
const {createReadStream} = require('fs');
const zlib = require('zlib');
const userGroupStateRepository = require('../repositories/userGroupStateRepository');
const groupRepository = require('../repositories/groupRepository');
const userRepository = require('../repositories/userRepository');
const unzipper = require('unzipper');
const pipelineAsync = promisify(pipeline);

class TestService {
  constructor() {
    this.userRepository = userRepository;
    this.groupRepository = groupRepository;
    this.userGroupStateRepository = userGroupStateRepository;
    this.tempFilesByOriginalFilename = {};
  }

  async handleFileUpload(file) {
    try {
      await this.unzipFile(file);
      await this.importFromFile();
    } finally {
      this.cleanupTempFiles();
    }
  }

  async unzipFile(file) {
    const tempDirPath = this.createTempDir();

    try {
      const extractionPromises = [];

      // Open ZIP file
      const filePath = file.path;
      const zipStream = createReadStream(filePath).pipe(unzipper.Parse({forceStream: true}));

      for await (const entry of zipStream) {
        if (entry.path.endsWith('.csv')) {
          const tempFilePath = path.join(tempDirPath, entry.path);
          const destination = fs.createWriteStream(tempFilePath);
          this.tempFilesByOriginalFilename[entry.path] = tempFilePath;

          // Pipe entry stream to file
          const extractionPromise = pipelineAsync(entry, destination)
            .catch(err => {
              console.error(`Error extracting ${entry.path}:`, err);
              throw new Error(`Error extracting ${entry.path}: ${err.message}`);
            });

          extractionPromises.push(extractionPromise);
        } else {
          entry.autodrain(); // Ignore non-CSV files
        }
      }

      await Promise.all(extractionPromises);

    } catch (err) {
      console.error('Error extracting ZIP file:', err);
      throw new Error(`Error extracting ZIP file: ${err.message}`);
    }
  }

  createTempDir() {
    const tempDir = os.tmpdir();
    const uniqueFolderName = `tmp_${uuidv4()}`;
    const tempDirPath = path.join(tempDir, uniqueFolderName);
    fs.mkdirSync(tempDirPath, {recursive: true});
    return tempDirPath;
  }

  getTempFile(fileName) {
    return this.tempFilesByOriginalFilename[fileName];
  }

  cleanupTempFiles() {
    for (const filePath of Object.values(this.tempFilesByOriginalFilename)) {
      try {
        fs.unlinkSync(filePath);
      } catch (e) {
        console.error(`Failed to delete temp file ${filePath}: ${e}`);
      }
    }
    // Delete also directories
    for (const dirPath of Object.values(this.tempFilesByOriginalFilename)) {
      if (fs.existsSync(dirPath)) {
        try {
          fs.rmdirSync(dirPath, {recursive: true});
        } catch (e) {
          console.error(`Failed to delete temp directory ${dirPath}: ${e}`);
        }
      }
    }
    this.tempFilesByOriginalFilename = {};
  }

  async importFromFile() {
    await this.insertGroupState();
    await this.insertGroup();
    await this.insertUser();
  }

  async insertGroupState() {
    const filePath = this.getTempFile('usergroupstate.csv');
    if (!filePath) return;

    const lines = fs.readFileSync(filePath, 'utf-8').split('\n').slice(1); // Skip header

    try {
      for (const line of lines) {
        if (line.trim() === '') continue;
        const [id, label] = line.split(',');
        const state = {
          id: parseInt(id),
          label: label,
        };
        await this.userGroupStateRepository.merge(state);
      }
    } catch (e) {
      console.error('Error inserting group states:', e);
      throw new Error(`Error inserting group states: ${e.message}`);
    }
  }

  async insertGroup() {
    const filePath = this.getTempFile('group.csv');
    if (!filePath) return;

    const lines = fs.readFileSync(filePath, 'utf-8').split('\n').slice(1); // Skip header

    try {
      for (const line of lines) {
        if (line.trim() === '') continue;
        const [
          id,
          name,
          contract_signed,
          contract_number,
          institute_name,
          institute_address_name,
          institute_address_number,
          institute_postal_code,
          province,
          city,
          country,
          vat,
          institute_phone,
          institute_email,
          state_id,
          priority,
          responsible_user_id
        ] = line.split(',');
        const group = {
          id: parseInt(id),
          name: name,
          contractSigned: new Date(contract_signed),
          contractNumber: parseInt(contract_number),
          instituteName: institute_name,
          instituteAddressName: institute_address_name,
          instituteAddressNumber: institute_address_number,
          institutePostalCode: institute_postal_code,
          province: province,
          city: city,
          country: country,
          vat: vat,
          institutePhone: institute_phone,
          instituteEmail: institute_email,
          stateId: parseInt(state_id),
          priority: parseFloat(priority),
        };
        if (id) {
          await this.groupRepository.merge(group);
        }
      }
    } catch (e) {
      console.error('Error inserting groups:', e);
      throw new Error(`Error inserting groups: ${e.message}`);
    }
  }

  async insertUser() {
    const filePath = this.getTempFile('user.csv');
    if (!filePath) return;

    const lines = fs.readFileSync(filePath, 'utf-8').split('\n').slice(1); // Skip header

    try {
      for (const line of lines) {
        if (line.trim() === '') continue;
        const [
          id,
          group_id,
          state_id,
          name,
          surname,
          login,
          passwd,
          email,
          phone,
          is_dsm_admin,
          is_group_admin,
          is_planner,
          is_support_tech,
          photo,
        ] = line.split(',');
        const user = {
          id: parseInt(id),
          groupId: parseInt(group_id),
          stateId: parseInt(state_id),
          name: name,
          surname: surname,
          login: login,
          passwd: passwd,
          email: email,
          phone: phone,
          isDsmAdmin: is_dsm_admin.toLowerCase() === 'true',
          isGroupAdmin: is_group_admin.toLowerCase() === 'true',
          isPlanner: is_planner.toLowerCase() === 'true',
          isSupportTech: is_support_tech.toLowerCase() === 'true',
          photo: photo,
        };
        if (id) {
          await this.userRepository.merge(user);
        }
      }
    } catch (e) {
      console.error('Error inserting users:', e);
      throw new Error(`Error inserting users: ${e.message}`);
    }
  }
}

module.exports = TestService;
